src/libserialize/leb128.rs

```diff
@@ -18,56 +20,95 @@ fn write_to_vec(vec: &mut Vec<u8>, position: usize, byte: u8) {
-    where W: FnMut(usize, u8)
-{
-    let mut position = 0;
-    loop {
-        let mut byte = (value & 0x7F) as u8;
-        value >>= 7;
-        if value != 0 {
-            byte |= 0x80;
-        }
-
-        write(position, byte);
-        position += 1;
+fn write_slice_to_vec(output: &mut Vec<u8>, start_position: usize, input: &[u8]) {
+    let input_len = input.len();
+    let capacity = output.len() - start_position;
+    let first_half = cmp::min(capacity, input_len);
 
-        if value == 0 {
-            break;
-        }
+    if first_half > 0 {
+        (&mut output[start_position..]).copy_from_slice(&input[.. first_half]);
     }
 
-    position
+    if first_half < input_len {
+        output.extend_from_slice(&input[first_half..]);
+    }
 }
 
-pub fn write_unsigned_leb128(out: &mut Vec<u8>, start_position: usize, value: u128) -> usize {
-    write_unsigned_leb128_to(value, |i, v| write_to_vec(out, start_position+i, v))
+#[cfg(target_pointer_width = "32")]
+const USIZE_LEB128_SIZE: usize = 5;
+#[cfg(target_pointer_width = "64")]
+const USIZE_LEB128_SIZE: usize = 10;
+
+macro_rules! leb128_size {
+    (u16) => (3);
+    (u32) => (5);
+    (u64) => (10);
+    (u128) => (19);
+    (usize) => (USIZE_LEB128_SIZE);
 }
 
-#[inline]
-pub fn read_unsigned_leb128(data: &[u8], start_position: usize) -> (u128, usize) {
-    let mut result = 0;
-    let mut shift = 0;
-    let mut position = start_position;
-    loop {
-        let byte = data[position];
-        position += 1;
-        result |= ((byte & 0x7F) as u128) << shift;
-        if (byte & 0x80) == 0 {
-            break;
+macro_rules! impl_write_unsigned_leb128 {
+    ($fn_name:ident, $int_ty:ident) => (
+        #[inline]
+        pub fn $fn_name(out: &mut Vec<u8>, start_position: usize, mut value: $int_ty) -> usize {
+            let mut encoded = [0u8; leb128_size!($int_ty)];
+
+            for i in 0 .. leb128_size!($int_ty) {
+                encoded[i] = (value as u8) & 0b0111_1111;
+                value = value >> 7;
+
+                if value == 0 {
+                    let bytes_written = i + 1;
+                    write_slice_to_vec(out, start_position, &encoded[0 .. bytes_written]);
+                    return bytes_written
+                } else {
+                    encoded[i] |= 0b1000_0000;
+                }
+            }
+
+            unreachable!()
         }
-        shift += 7;
-    }
+    )
+}
 
-    (result, position - start_position)
+impl_write_unsigned_leb128!(write_u16_leb128, u16);
+impl_write_unsigned_leb128!(write_u32_leb128, u32);
+impl_write_unsigned_leb128!(write_u64_leb128, u64);
+impl_write_unsigned_leb128!(write_u128_leb128, u128);
+impl_write_unsigned_leb128!(write_usize_leb128, usize);
+
+macro_rules! impl_read_unsigned_leb128 {
+    ($fn_name:ident, $int_ty:ident) => (
+        #[inline]
+        pub fn $fn_name(data: &[u8], start_position: usize) -> ($int_ty, usize) {
```

Comment:
This was here before, bu t why does this take a slice and an offset instead of just a slice?

Reply:
Good question. Maybe there was a reason in the past. I'll change it.